### PR TITLE
Update major interfaces

### PR DESCRIPTION
**Pull request by junha1**
This PR includes partial impl. of some interface. Please take a look at it to understand the new interface
Also, the PR consists of multiple commits. Review each commit to see the updates clearly.

### CODE REVIEW

File: network/src/lib.rs [hunk 1]
```diff
@@ -10,15 +10,19 @@ use std::{net::SocketAddrV4, sync::Arc};
...
 pub enum Error {
     #[error("unknown error: {0}")]
     Unknown(String),
 }
 
+/// The information of a network peer that is discovered by the discovery protocol.
 #[derive(Debug, PartialEq, Eq, Clone, Serialize, Deserialize)]
 pub struct Peer {
     pub public_key: PublicKey,
+    /// The address used for the discovery protoocl
```
Comment (TomTaehoonKim) [inline network/src/lib.rs:25]:
```suggestion
    /// The address used for the discovery protocol
```

File: repository/src/lib.rs [hunk 2]
```diff
@@ -9,3 +15,118 @@ pub type Tag = String;
...
+        self,
+        network_config: &NetworkConfig,
+        peers: SharedKnownPeers,
+    ) -> Result<tokio::task::JoinHandle<Result<(), Error>>, Error>;
+
+    /// Checks the validity of the repository, starting from the given height.
+    ///
+    /// It checks
+    /// 1. all the reserved branches and tags
+    /// 2. existence of merge commits
```
Comment (TomTaehoonKim) [inline repository/src/lib.rs:82]:
```suggestion
    /// 2. the existence of merge commits
```

File: repository/src/lib.rs [hunk 3]
```diff
@@ -9,3 +15,118 @@ pub type Tag = String;
...
+    /// This will verify every commit along the way.
+    /// If the given commit is not a descendant of the
+    /// current `main` (i.e., cannot be fast-forwarded), it fails.
+    ///
+    /// Note that if you sync to a block `H`, then the `main` branch will move to `H-1`.
+    /// To sync the last block `H`, you have to run `finalize()`.
+    /// (This is because the finalization proof for a block appears in the next block.)
+    async fn sync(&mut self, block_commit: &CommitHash) -> Result<(), Error>;
+
+    /// Returns the current valid and height-acceptable agendas in the repository.
```
Comment (TomTaehoonKim) [inline repository/src/lib.rs:97]:
```suggestion
    /// Returns the currently valid and height-acceptable agendas in the repository.
```

File: repository/src/lib.rs [hunk 4]
```diff
@@ -9,3 +15,118 @@ pub type Tag = String;
...
+    async fn get_agendas(&self) -> Result<Vec<(CommitHash, Hash256)>, Error>;
+
+    /// Returns the current valid and height-acceptable blocks in the repository.
```
Comment (TomTaehoonKim) [inline repository/src/lib.rs:100]:
```suggestion
    /// Returns the currently valid and height-acceptable agendas in the repository.
    async fn get_agendas(&self) -> Result<Vec<(CommitHash, Hash256)>, Error>;

    /// Returns the currently valid and height-acceptable blocks in the repository.
```

File: consensus/src/lib.rs [hunk 5]
```diff
@@ -34,29 +34,33 @@ pub enum ProgressResult {
...
     async fn veto_block(
-        directory: &str,
+        &mut self,
         network_config: NetworkConfig,
         known_peers: &[Peer],
         block_hash: Hash256,
-        height_to_assert: BlockHeight,
     ) -> Result<(), Error>;
 
+    async fn set_proposal(&mut self, block_hash: Hash256) -> Result<(), Error>;
```
Comment (lemonholic) [inline consensus/src/lib.rs:57]: What this function does? Marking a block to be proposed when `progress` is called?

File: node/src/lib.rs [hunk 6]
```diff
@@ -1,245 +1,99 @@
...
-    pub state_transition: Option<StateTransition>,
-    /// An optional field to store arbitrary data,
-    /// which is not part of the state but still useful as it's still part of the finalizing blockchain data
-    /// and so can be verified with the Merkle root.
-    ///
-    /// Note that it must be `Some` if the `state_transition` is `None`.
-    /// Otherwise, the transaction would be meaningless.
-    pub payload: Option<String>,
-}
+/// It is for serving the **CLI**, providing low-level funcitons and type-specified interfaces.
```
Comment (lemonholic) [inline node/src/lib.rs:40]: funcitons

File: node/src/lib.rs [hunk 7]
```diff
@@ -1,245 +1,99 @@
...
-    ReceiveFinalizedBlock {
-        block: Block,
-        finalization_proof: FinalizationProof,
-        timestamp: Timestamp,
-    },
-    Timer {
-        timestamp: Timestamp,
-    },
-}
+    /// Cleans the repository, removing all the oudated commits.
```
Comment (lemonholic) [inline node/src/lib.rs:56]: oudated

File: node/src/lib.rs [hunk 8]
```diff
@@ -1,245 +1,99 @@
...
-    /// Otherwise, the transaction would be meaningless.
-    pub payload: Option<String>,
-}
+/// It is for serving the **CLI**, providing low-level funcitons and type-specified interfaces.
+#[async_trait]
+pub trait SimperbyApi<RR: RawRepository, R: DistributedRepository<RR>, C: Consensus, G: Governance>
+{
+    /// Initializes a new Simperby node from the genesis state
+    /// stored in the given directory (which is not yet a Git repository).
+    async fn genesis(&self) -> Result<()>;
```
Comment (lemonholic) [inline node/src/lib.rs:46]: Will you use `Config::repository_directory` as the directory for genesis info?

File: repository/src/lib.rs [hunk 9]
```diff
@@ -9,3 +15,118 @@ pub type Tag = String;
...
+
+    /// Returns the block header from the `main` branch.
+    async fn get_last_finalized_block_header(&self) -> Result<BlockHeader, Error>;
+
+    /// Fetches new commits from the network.
+    /// It **verifies** all the incoming changes and applies them to the local repository
+    /// only if they are valid.
+    async fn fetch(
+        &mut self,
+        network_config: NetworkConfig,
```
Comment (lemonholic) [inline repository/src/lib.rs:58]: `&NetworkConfig`? (by the previous style)

File: repository/src/lib.rs [hunk 10]
```diff
@@ -9,3 +15,118 @@ pub type Tag = String;
...
+        known_peers: &[Peer],
+    ) -> Result<(), Error>;
+
+    /// Notifies there was a push for the given repository.
+    async fn notify_push(
+        &mut self,
+        network_config: NetworkConfig,
```
Comment (lemonholic) [inline repository/src/lib.rs:65]: ditto

File: repository/src/lib.rs [hunk 11]
```diff
@@ -9,3 +15,118 @@ pub type Tag = String;
...
+    ///
+    /// It will verify the finalization proof and the commits.
+    async fn finalize(
+        &mut self,
+        block_commit_hash: &CommitHash,
+        proof: &FinalizationProof,
+    ) -> Result<(), Error>;
+
+    /// Informs that the given agenda has been approved.
+    async fn approve(&mut self, agenda_commit_hash: &CommitHash) -> Result<(), Error>;
```
Comment (lemonholic) [inline repository/src/lib.rs:113]: Will it be called by `governance` module and then create an empty `agenda-proof` commit?

File: repository/src/lib.rs [hunk 12]
```diff
@@ -9,3 +15,118 @@ pub type Tag = String;
 pub struct CommitHash {
     pub hash: [u8; 32],
```
Comment (JeongHunP) [inline repository/src/lib.rs:16]: Isn't this [u8; 20]? CommitHash consists of 40 hex characters

Comment (TomTaehoonKim) [reply repository/src/lib.rs:16]: It’s 256 bits for the hash value.

File: consensus/src/lib.rs (hunk 5 above)
Comment (junha1) [reply consensus/src/lib.rs:57]: yes.

File: repository/src/lib.rs (hunk 4 above)
Comment (junha1) [reply repository/src/lib.rs:100]: it's (current), (valid), not (current valid)

File: consensus/src/lib.rs [hunk 13]
```diff
@@ -34,29 +34,33 @@ pub enum ProgressResult {
 }
 
 #[async_trait]
-pub trait Consensus {
+pub trait Consensus: Send + Sync {
```
[the remark this answers is not in the document]
Comment (junha1) [reply consensus/src/lib.rs:37]: Because of async programming. This is really common, FYI.

File: node/src/lib.rs (hunk 8 above)
Comment (junha1) [reply node/src/lib.rs:46]: Yes, which is implied in the comment `which is not yet a Git repository`

File: repository/src/lib.rs (hunk 11 above)
Comment (junha1) [reply repository/src/lib.rs:113]: No. It is called by `Node`. There are no dependencies between governance, consensus, chatting, and repo.
Node coordinates everything.

File: repository/src/lib.rs [hunk 14]
```diff
@@ -9,3 +15,118 @@ pub type Tag = String;
...
+}
+
+/// The local Simperby blockchain data repository.
+///
+/// It automatically locks the repository once created.
+///
+/// - It **verifies** all the incoming changes and applies them to the local repository
+/// only if they are valid.
+#[async_trait]
+pub trait DistributedRepository<T: RawRepository>: Send + Sync {
```
[the remark this answers is not in the document]
Comment (junha1) [reply repository/src/lib.rs:42]: RR is a mere wrapper of Git2, whereas DR is about the actual Simperby protocol. (verification, commit generation, high-level operations)

File: repository/src/lib.rs (hunk 12 above)
Comment (junha1) [reply repository/src/lib.rs]: @JeongHunP You are right. I fixed it.

File: consensus/src/lib.rs (hunk 5 above)
Comment (lemonholic) [reply consensus/src/lib.rs:57]: thanks

File: consensus/src/lib.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (lemonholic) [reply consensus/src/lib.rs:37]: ok

File: repository/src/lib.rs (hunk 11 above)
Comment (lemonholic) [reply repository/src/lib.rs:113]: I see.

Comment (lemonholic) [reply repository/src/lib.rs:113]: Then did I understand correctly with "create an empty agenda-proof commit" ?

File: node/src/lib.rs (hunk 8 above)
Comment (lemonholic) [reply node/src/lib.rs:46]: Ok

File: repository/src/lib.rs (hunk 11 above)
Comment (junha1) [reply repository/src/lib.rs]: No. I was wrong. I add a fix commit.

Comment (lemonholic) [reply repository/src/lib.rs]: checked it

**Review by lemonholic — APPROVED** [review]
good to go

**Pull request merged**